### PR TITLE
[feat] Add formatting to markdown link text

### PR DESCRIPTION
**Pull request by mdneyazahmad**
@parasharrajat  will you please review this?

This PR tries to add a feature to apply selected formatting rules in the markdown parser and uses this feature to format the link text.

`ExpensiMark` replace method now accepts an optional argument.

```
    /**
     * Replaces markdown with html elements
     *
     * @param {String} text - Text to parse as markdown
     * @param {Object} [options] - Options to customize the markdown parser
     * @param {String[]} [options.filterRules=[]] - An array of name of rules as defined in this class.
     * If not provided, all available rules will be applied.
     * @param {Boolean} [options.isTextEscaped=false] - Whether the text is already escaped
     *
     * @returns {String}
     */
```

### Fixed Issues
$ https://github.com/Expensify/App/issues/8426

### Related PR
$ https://github.com/Expensify/App/pull/9022

# Tests
1. Open the app
2. Select a report
3. Send `[~Hi~ **Hello** _Expensify_](expensify.com)`
4. Verify that `~Hi~ **Hello** _Expensify_` is formatted as per the rules.

# QA
1. Open the app
2. Select a report
3. Send `[~Hi~ **Hello** _Expensify_](expensify.com)`
4. Verify that `~Hi~ **Hello** _Expensify_` is formatted as per the rules.

# Screenshots
### Web
web

### Desktop
desktop

### Mobile web (Safari)
mweb

### iOS
ios

### Android
android

### CODE REVIEW

**Comment by mdneyazahmad** [comment]
I have read the CLA Document and I hereby sign the CLA

File: lib/ExpensiMark.js [hunk 1]
```diff
@@ -367,7 +370,9 @@ export default class ExpensiMark {
             if (abort) {
                 replacedText = replacedText.concat(textToCheck.substr(match.index, (match[0].length)));
             } else {
-                replacedText = replacedText.concat(replacement(match[0], match[1], match[2], match[3]));
+                // we are allowing only bold, strikethrough and italic formatting for link text
```
Comment (parasharrajat) [inline lib/ExpensiMark.js:373]: Comment can be improved. Mention why are we doing this. Comment starts from capitals.

File: lib/ExpensiMark.js [hunk 2]
```diff
@@ -291,14 +292,16 @@ export default class ExpensiMark {
      * Replaces markdown with html elements
      *
      * @param {String} text
+     * @param {String[]} rules
      *
      * @returns {String}
      */
-    replace(text) {
+    replace(text, rules = []) {
```
Comment (parasharrajat) [inline lib/ExpensiMark.js:299]:
```suggestion
    replace(text, filterRules = []) {
```
What rules? define the param in a comment.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Add more unit tests.

**Comment by mdneyazahmad** [comment]
@parasharrajat updated

File: __tests__/ExpensiMark-Markdown-test.js [hunk 3]
```diff
@@ -158,6 +161,9 @@ test('Test acnchor tags convesion to markdown style link with various styles', (
         + '[Expensify!](https://www.expensify.com) '
         + '[Expensify?](https://www.expensify.com) '
         + '[Expensify](https://www.expensify-test.com) '
+        + '[*Expensify*](https://www.expensify-test.com) '
+        + '[test.com](https://www.expensify-test.com) '
+        + '[_italic_ ~strikethrough~ test.com](https://www.expensify-test.com) '
```
[the remark this answers is not in the document]
Comment (mdneyazahmad) [reply __tests__/ExpensiMark-Markdown-test.js:166]: It passes this test. However the reverse is not producing the exact code.

---

`[https://www.text.com/_root_folder/1](https://www.text.com/_root_folder/1)` 
To
`<a href="https://www.text.com/_root_folder/1" target="_blank" rel="noreferrer noopener">https://www.text.com/_root_folder/1</a>`

---

`<a href="https://www.text.com/_root_folder/1" target="_blank">https://www.text.com/_root_folder/1</a>`
To
`https://www.text.com/_root_folder/1` and not `[https://www.text.com/_root_folder/1](https://www.text.com/_root_folder/1)` 

---

I think, it is fine. That is the same thing.

**Comment by mdneyazahmad** [comment]
Updated

**Review by MonilBhavsar — APPROVED** [review]
Looks good to me!

**Comment by MonilBhavsar** [comment]
In the related PR, in the QA section, let's add to test other markdown styles too

**Comment by MonilBhavsar** [comment]
Feel free to update your other PR with commit hash `e989d858aa2ab8191284ae106acf9cd323ab879d`

**Pull request merged**